lexer/stateful/stateful.go

```diff
@@ -64,8 +64,147 @@ type compiledRule struct {
+	// big to put in the runeMap (like '.') and that are tested all the time.
+	runeMap     map[rune][]bool
+	fallbackMap []bool
+}
+
+func (group *compiledRuleGroup) process() error {
+	var (
+		rules    = group.rules
+		nbrules  = len(rules)
+		computed = make([]*computedRuneRange, nbrules)
+		fallback = make([]bool, nbrules)
+		runemap  = make(map[rune][]bool)
+	)
+	group.runeMap = runemap
+	group.fallbackMap = fallback
+
+	// Analyze the patterns to populate runeMap and fallbackMap
+	for i, rule := range rules {
+		if rule.RE == nil || rule.Pattern == "" {
+			fallback[i] = true
+			continue // ignore those
+		}
+		comp, err := computeRuneRanges(rule.Pattern)
+		if err != nil {
+			return err // FIXME better error handling ?
+		}
+		computed[i] = comp
+		if comp.size >= charclassSizeLimit {
+			// this one is too big to have their own map
+			fallback[i] = true
+			continue
+		}
+	}
+
+	// Now process the rules and add them to their respective buckets.
+	for idrule, comp := range computed {
+		if comp == nil || fallback[idrule] {
+			continue
+		}
+		var runes = comp.runes
+		for i, l := 0, len(runes); i < l; i += 2 {
+			for r, end := runes[i], runes[i+1]; r <= end; r++ {
+				var (
+					bools []bool
+					ok    bool
+				)
+				if bools, ok = runemap[r]; !ok {
+					bools = make([]bool, nbrules)
+					runemap[r] = bools
+					// when creating a new bucket, add all the fallback rules onto it if they could potentially
+					for j, isFallback := range fallback {
+						if isFallback {
+							var comp = computed[j]
+							// add all the fallback rules that are either not yet defined (or "")
+							// or those that we know will never match this rune.
+							bools[j] = comp == nil || !runeIsInRange(r, comp)
+						}
+					}
+				}
+
+				bools[idrule] = true // add it onto the bucket.
+			}
+		}
+	}
+
+	return nil
+}
+
+func runeIsInRange(r rune, comp *computedRuneRange) bool {
+	if comp == nil {
+		return true // all the unknown rules are added.
+	}
+	var rng = comp.runes
+	for i, l := 0, len(rng); i < l; i += 2 {
+		if r >= rng[i] && r <= rng[i+1] {
+			return true
+		}
+	}
+	return false
+}
+
+// Tries to match to input. Returns nil, nil if no match was found.
+func (group *compiledRuleGroup) tryMatch(l *Lexer) (*compiledRule, []int, error) {
+	// Our goal is to find the indices of the rules to test
+	var (
+		indices []bool
+		ok      bool
+		rules   = group.rules
+		re      *regexp.Regexp
+		err     error
+	)
+
+	// first, get the first rune
+	r, _ := utf8.DecodeRune(l.data)
+	if r == utf8.RuneError {
+		return nil, nil, fmt.Errorf(`invalid utf-8 input sequence`) // FIXME this should be an error, how do we report it ?
+	}
+
+	if indices, ok = group.runeMap[r]; !ok {
+		indices = group.fallbackMap // so we use the fallback to test the rest.
+	}
+
+	for i, test := range indices {
+		if !test {
+			continue
+		}
+
+		var rule = &rules[i]
+
+		// FIXME: what about one char rules ?
```

Comment:
Do these FIXME's matter? ie. will user's encounter them or not?

If they will, they should be fixed now, as (generally) FIXME's are never fixed. Ironic.